File: 841-shortest-distance-to-a-character/shortest-distance-to-a-character.py
class Solution:
    def shortestToChar(self, s, c):
        n = len(s)
        ans = [n] * n

        prev = -n

        for i in range(n):
            if s[i] == c:
                prev = i

            ans[i] = i - prev

        prev = n * 2

        for i in range(n - 1, -1, -1):
            if s[i] == c:
                prev = i

            ans[i] = min(ans[i], prev - i)

        return ans

File: 841-shortest-distance-to-a-character/shortest-distance-to-a-character.java
class Solution {
    public int[] shortestToChar(String s, char c) {

        int n = s.length();
        int[] ans = new int[n];

        int prev = -n;

        for (int i = 0; i < n; i++) {

            if (s.charAt(i) == c) {
                prev = i;
            }

            ans[i] = i - prev;
        }

        prev = n * 2;

        for (int i = n - 1; i >= 0; i--) {

            if (s.charAt(i) == c) {
                prev = i;
            }

            ans[i] = Math.min(
                ans[i],
                prev - i
            );
        }

        return ans;
    }
}